### PR TITLE
Implement CPU backward convolution functions in extern C

### PR DESCRIPTION
**Pull request by yiqian1**
Added two cpu backward convolution functions: mcpuConv2dBwdData() and mcpuConv2dBwdWeight(). Thus all three cpu convolutions are implemented separately. 

I've attempted to reuse mcpuConv2d() for bwd_weight. It was doable. However it required manipulating dimensions of all tensors, and made the code a bit complicated and less readable. So I've decided to implement convolution operations in separate functions which are much cleaner. 

I've run 98 backward convolution test cases, and got 90 result failures, 8 crashes, and 0 passes. Failed tests are filed in [issue#70](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/70) and [#71](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/71).

### CODE REVIEW

**Review by asroy — APPROVED** [review]
LGTM

File: mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp [hunk 1]
```diff
@@ -440,3 +457,119 @@ extern "C" void mcpuConv2d(int64_t rank1, void *f_ptr, int64_t rank2,
...
+  auto *filterAllocated = filter->data + filter->offset;
+
+  auto *input = static_cast<StridedMemRefType<float, 4> *>(i_ptr);
+  auto *inputAllocated = input->data + input->offset;
+
+  auto *output = static_cast<StridedMemRefType<float, 4> *>(o_ptr);
+  auto *outputAllocated = output->data + output->offset;
+
+  // Extract proper tensor sizes and strides based on layouts
+  TensorDim filterSizeStride, inputSizeStride, outputSizeStride;
```
Comment (jerryyin) [inline mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp:482]: Can you help me understand something: The whole calculation below is highly dependent on stride and layout correlated with each other. This means that for a memref with a different layout, the strides must be different (For example, NCHW and NHWC have different strides). However, we have never altered the strides anywhere according to layout at all. How can we make sure that the strides we obtained here is related with its layout?

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 2]
```diff
@@ -1087,40 +1093,74 @@ static LogicalResult populateValidationLogic(ModuleOp &module,
...
+        builder.getUnknownLoc(), mcpuMemset4DFuncOp,
+        ValueRange{cpuOutputMemRefCastOp, outputMemsetValue});
+    block->push_back(cpuOutputMemsetOp);
 
-  // Emit memref cast
-  auto cpuOutputMemRefCastOp = builder.create<MemRefCastOp>(
-      builder.getUnknownLoc(), cpuOutputHostAllocOp,
-      fourDimUnknownSizeMemRefType);
-  block->push_back(cpuOutputMemRefCastOp);
+  } else if (operation.getValue() == "conv2d_bwd_data") {
```
Comment (jerryyin) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: Could you consolidate this long if --- else if --- else if --- routine into a common function here. I figure there's mostly duplicated code among three blocks of code.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 3]
```diff
@@ -442,9 +442,19 @@ static FuncOp createCPUConvolution(ModuleOp &module, OpBuilder &builder,
   cpuConvBlock->push_back(inLayoutMemRefCastOp);
```
Comment (jerryyin) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:470]: This is a note to remember that the existing validation logic (based on whether cpu/gpu results are equal, and input of 1) is not good enough. To correctly validate the cpu/gpu results. There needs to be input of random numbers. Also, to compare float number close enough based on subtraction. 

If you decide to address that in a new review/issue I'd be fine with that.

File: mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp (hunk 1 above)
Comment (yiqian1) [reply mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp:482]: This is handled automatically by API. No matter how the layout changes, sizes and strides are always aligned with its corresponding layout. That means the stride of 'C' in the StridedMemRefType of NCHW is different than the stride of 'C" in the StridedMemRefType of NHWC. What I need to do is just find the right strides/sizes based on layouts.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp (hunk 3 above)
Comment (yiqian1) [reply mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:470]: Agree. We should test on random inputs.

File: mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp (hunk 1 above)
Comment (jerryyin) [reply mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp:482]: Ok, it looks like if it is a NCHW tensor, then we will always allocate a memref<N, C, H, W>; whereas if it is NHWC tensor, we always allocate a memref<N, H, W, C>. By allocating memref differently, we can obtain strides correctly according to that layout.

**Review by jerryyin — APPROVED** [review]

**Pull request merged**